### PR TITLE
Add datafusion example of expression apis

### PR DESCRIPTION
**Pull request by alamb**
~Draft as it builds on:~
- [x] https://github.com/apache/arrow-datafusion/pull/3719
- [x] ~https://github.com/apache/arrow-datafusion/pull/3742~
- [x] https://github.com/apache/arrow-datafusion/pull/3758

# Which issue does this PR close?
re https://github.com/apache/arrow-datafusion/pull/3719
re https://github.com/apache/arrow-datafusion/issues/3708
re https://github.com/apache/arrow-datafusion/issues/3740

# Rationale for this change
Documenting the APIs will make the more discoverable and improve the user experience

# What changes are included in this PR?
Add a new `expr_api` example

# Are there any user-facing changes?
Better docs

### CODE REVIEW

**Review by andygrove — APPROVED** [review]
This is great. Thanks @alamb

**Comment by alamb** [comment]
I moved my proposed API changes into https://github.com/apache/arrow-datafusion/pull/3758

File: datafusion-examples/examples/expr_api.rs [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+    // basic arithmetic simplification
+    // i + 1 + 2 => a + 3
+    // (note this is not done if the expr is (col("i") + (lit(1) + lit(2))))
+    assert_eq!(
+        simplifier.simplify(col("i") + (lit(1) + lit(2)))?,
+        col("i") + lit(3)
+    );
+
+    // TODO uncomment when https://github.com/apache/arrow-datafusion/issues/1160 is done
```
Comment (alamb) [inline datafusion-examples/examples/expr_api.rs:109]: I will do a partial fix for this in a follow on PR

File: datafusion/core/src/prelude.rs [hunk 2]
```diff
@@ -34,7 +34,7 @@ pub use crate::execution::options::{
 pub use datafusion_common::Column;
 pub use datafusion_expr::{
     expr_fn::*,
-    lit,
+    lit, lit_timestamp_nano,
```
Comment (alamb) [inline datafusion/core/src/prelude.rs:37]: driveby fix
